### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] Option to ignore arrays

### ISSUE THREAD

**Issue by astorije** [typescript-eslint/typescript-eslint#1798]
Consider the following snippet:

```ts
const foo = ['a', 'b', 'c']; // type: string[]
const bar = foo[5]; // type: string, value: undefined

// Passes the rule
if (foo[5]) {
  console.log(foo[5]);
}

// Fails the rule
if (bar) {
  console.log(bar);
}
```

Thanks to @Retsam's work on https://github.com/typescript-eslint/typescript-eslint/pull/1534, the first example passes the linter, however as the author mentions in their PR, it doesn't handle less trivial cases, such as storing the value in a variable (which is very common when you want to use the value being tested in the body of the `if`).

I realize that supporting this (and other advanced) cases might be a lift. In the meantime, or alternatively, any thoughts on an option to entirely ignore arrays? In the current state, I don't believe there is a way to make the second example above pass the linter without disabling the rule or getting rid of the variable.
I believe such option was mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/1544, for reference.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.24.0` |
| `@typescript-eslint/parser`        | `2.24.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.0` |
| `npm`                              | `6.13.4` |

**Comment by bradzacher** [comment]
It would not really be possible, no.

From the POV of typescript, a string is a string. Doesn't matter where it came from.
So for TS, there's no difference between `const x = 'a'` and `const y = foo[5]` - they both just have a string type.

Which means in order to detect this, the rule would have to do control-flow analysis of a variable to determine where the value in any variable has come from.
In some cases this might be simple to do, but it very, very quickly degrades into an intractable problem to solve (see some examples below).

There's but one case that's "easy" to solve, and that is a const variable whose value was directly assigned from an array index:
```ts
const bar = foo[5];
if (bar) {}
```

Happy to accept a PR if you want to support this case, otherwise I'd say that it's not possible to do.

----

```ts
const foo = ['a', 'b', 'c'];
const bar = foo[5];

if (bar) {} // "relatively" easy to detect, as there has been exactly one assignment.

let bam = foo[5];

bam = 'a';
if (bam) {} // hard to tell that if the value is an array

if (someCondition) { bam = 'a'; }
if (bam) {} // impossible to know if the value came from an array

const buz = foo[5];
const bang = buz;
if (bang) {} // really, really, really hard to know if the value came from an array

test(bar);
function test(baz: string) {
  if (baz) {} // impossible to detect
}
 
```

**Comment by astorije** [comment]
Thanks @bradzacher, yet another amazing piece of feedback, very helpful :)
It's very interesting to learn more about the dependence on the language, the limitation or capabilities of various things, etc.

In the examples above, this is also an acceptable alternative and satisfies the linter:

```ts
const foo = ['a', 'b', 'c']; // type: string[]

// Passes the rule
if (5 in foo) {
  console.log(foo[5]);
}
```

(Or course it's a little silly like this but imagine a dynamic index)

As we go through all our projects one by one, we'll see if other cases arise. Should the format described above come up too often, I'll see if I can look at improving the linter.

**Comment by astorije** [comment]
Following-up on this, it would also be great to be able to ignore `Record`s as well. We are having the following case:

In our own code, we have something of the sort:

```tsx
if (!client.getResolvers().Foo?.id) {
    client.addResolvers({
      Foo: {
        id: /* ... */,
      },
    });
}
```

`client` is controlled by a third-party library, which defines `getResolvers()` return type as:

```tsx
[key: string]: {
  [field: string]: { /* ... */ };
};
```

In other parts of the code, we have been able to rewrite the value of the record as `[field: string]: { /* ... */ } | undefined;` but since this is a third-party dependency, we do not have this luxury here.

Because `client.getResolvers().Foo` is non-nullable type-wise, the rule flags the extra `?` as unnecessary, which is not true.

I think having one or multiple options to ignore things like this, lookups that can be undefined even though TS typing indicates it is not, would make things safer, as following the linter recommendation breaks the code here. What do you think?

**Comment by bradzacher** [comment]
The problem is that that that option would add a lot of unsafety in the rule, as you'll no longer get warned about records within your control.
I don't think that makes it a good option.

A workaround is to inline the call and apply a custom type:
```ts
declare const client: {
    getResolvers(): Record<string, Record<string, number>>;
    addResolvers(resolver: Record<string, Record<string, number>>): void;
};

type SafeResolversRecord = Record<string, Record<string, number | undefined> | undefined>;

const resolvers: SafeResolversRecord = client.getResolvers();
if (resolvers.Foo?.id) {
    client.addResolvers({
        Foo: {
            id: 1,
        },
    });
}
```

depending on how the module defines its types, you could even augment this type into the module itself to make it easier to consume:

```ts
import "foo";
declare module "foo" {
  export type SafeResolversRecord = Record<string, Record<string, number | undefined> | undefined>;
}

////////

import { SafeResolversRecord } from 'foo';
```

**Comment by astorije** [comment]
That's really great advice, thank you @bradzacher!

**Comment by sindresorhus** [comment]
https://github.com/microsoft/TypeScript/issues/13778 would fix this, right?

**Comment by bradzacher** [comment]
Yup! It would definitely fix this.
This is a problem for all index access in TS, and is a huge source of unsafety in codebases.
TBF - flow (which prioritises type soundness above all else) has this same problem. In this case, both systems just prioritise the devx over safety for this.

Because of #1534, we handle the trivial to detect case of `if (arr[0]) {}` (the rule detects that `arr` is an array type, and ignores the condition on it).
But it doesn't currently handle any more cases (like assigning to a variable).

We can use scope analysis to handle this case:
```ts
const x = arr[0];
if (x) {} // detectable
```
Any other cases mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/1798#issuecomment-604021798 get very fuzzy and likely to false positive/negative.

**Comment by jmicco** [comment]
In addition to array types, this also causes problems for Map index if I have a type like:
```
export interface ITargetEntry {
    repo: string
    path: string
}
export interface ITargetMap {[target: string]: ITargetEntry}
const targetMap: ITargetMap = {...}
const targetEntry: ITargetEntry = targetMap[product]
if (targetEntry) {
...
}
```
The index into the map can certainly be undefined - but the no-unnecessary-condition flags it - this is not an array index it is an index into a string map.

**Comment by Retsam** [comment]
@jmicco Yeah, that's a well-known limitation as well.  Unlike the array case, that one has a pretty good workaround, though.  It's better to declare your map in a way that indicates that keys might not exist:

```ts
interface ITargetMap {
    [target: string]: ITargetEntry | undefined
}
// Or `type ITargetMap = Partial<Record<string, ITargetEntry>>` for short
```

**Comment by jmicco** [comment]
Hi @retsam, I thought I tried that and it did not work - my apologies, that
fix did prevent the complaint, although it seems like the type imputation
for such a hash needs to generate the type as "x | undefined", but I
understand that is the problem for the typescript people and not for this
rule.

**Comment by publicvirtualvoid** [comment]
I really enjoy this rule because it makes tidying up after a refactor a lot more effective, but the situation at the moment is a little tricky to manage. 

`noUncheckedIndexedAccess` is not particularly smart and doesn't understand simple `for` loops. It can't be disabled per file/line like a lint rule, and enabling it on a large existing project can produce a huge number of build errors. Putting the necessary guards in place to resolve the errors would make some code considerably more convoluted.

Could something like `noUncheckedIndexedAccess` theoretically exist as an eslint rule such that it can be disabled in certain contexts?

**Comment by bradzacher** [comment]
theoretically - yes - it could be implemented.
practically - in order to implement it, you would have to do a bunch of work to rebuild everything TS does internally. It would be pretty complicated and a lot of work because you need to implement control-flow analysis as well so that you can understand like:
```ts
const x: number[] = [];
if (x[0]) {
  x[0].length; // is "safe" due to the condition 
}
```

Which is really the most complicated part of it. You'd ultimately just be rebuilding a chunk of TS

**Comment by ddhorstman** [comment]
I'd be interested in updating this rule to cover the most trivial additional case, as I've run into this issue in my own codebase:
```
const x = arr[0];
if (x) {} // detectable
```
However, I've never worked on linter rules before, and I'm not sure where to start. It sounds like this is detectable via "scope analysis" - could anyone offer advice on how I might be able to learn how to do this?

**Comment by bradzacher** [comment]
@ddhorstman - [I believe we already have handling for this simple case](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gTwFzyQCMArcDAbgCgBLAM3gApsBKeAb3gF9KB6H+EDBhIYaeIXBQArmgTZ41MQGsUSAO4p4GJPBQgAboPEIUUiBEqVQkWAmTossGPiKkwGANoBdKvczwAD3gAXngnDwAGHxp6BgDWDm4+eGAkEDFVLEFhGGMwaVlAxCQAW3T4WmESsM0nKGwgA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwo2V4A8pTIB7BogxhQECDHgiAVsmI5JAbR4yoOEmIBmZAOYA6BYjGRNYALrctkcwFd4xZABUCySZAC2IgCaOtF48AL52vLSOhowS6GqQuJ4k5JQ0iPRMzHSM1NBRMQyQtjyQ8EFEGNIyifiEKRTUOVkMIlSODAzKRIgIeFTEYv5kOKJFmJCo8mhhIKFAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by frontsideair** [comment]
I'm having trouble with false negatives and this is the only issue I could find on this topic. I can open a separate one if needed.

[Here's the repro](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3lMGCWqAIlBlAFzwDeAUPHfCALZR4QWYx4oDm8APvA5duAbQC6Ablr0ADlDRoA7khjB2GTj36CNw8VIC+2gK4pQAMy4hg1aiAAeMlVmToscNMYhYAvNlwEKMSkAPwAdEwsEOEiAAySQA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCUAAgE7QC2AlrALTECuSA9NGgDYVLzUDGA9kgDMKAc0gAafBEglyVWg2ZsOXPoJGMkvFNHGSoAB1Lx4FaMV0EDrOsI6ZC8AJ76W3YhX1cW7To1J8yMmRtFAsCSH1rWyR7Jxc0Nw8vJV9-XkDg6BRaaABHOgp3JGFqOOgeAAtobgBrDlEJSwiouwdnV3dPam9lRjR4d254SHwAXUaDAENiNDMsIjKEzuSfeEZ9adnzCYjNswB5Twp+DGxQJsNeACtq4ew8SykAOkZ4BP4hYSertH4Ry1G+AAvjtIjYOKdcHppItEl0epx-mBxvhIMReLw7mABnRoDt6KwWPNzuFNNQAO7TJD1HjpIKcSEPR4wYjo7Z6CAkx5QRDEMiMjnMoSwIJI7ko7lSVi8biTEx-bCQSZIRzkqqkMUEIF6CXhNrxOErXpkgBGk1mpV41H6RVEirMbLCUn1HSS3RSa1I+UK5UmlIoWMglOISCdC3aSzdCLWDBNvAY2SC8AqWnmkF4AgEYZhEcN7tWGityfR5Oo7F5k1YaYzWYmzthy3zvVUaAo-WQXDK1G0Qmp8qQAu5QapYakZU1uvrucb0cL1AYSGqLDQ00ctNQAeOocVwdDwJ2Ik0pAACnLeQP5kypIXtAB9MhaOiEjB1qDMABMOlfkEYk30FGzRgTQKVhQm-dY6BNdhuCAkCwOhAAqBDGAQ74MDGEAgSAA&tsconfig=N4KABGBECWB2DGAbArgEwKaQFxgNrgigCd0BbaADwDp1YA3K1KgFwGdIAaAiSAKl4D0vFuy6Eo-ISIqQCAXTFR4Ae1IAHaInREA8mubRlsdjlDjIiaACNseboUipVnezyekq0ZtoCGVrS7iPOisAEwADACMAJyy4gr2kCEAssqoyFoAkrDeRMpqtsxEyOiKPABWrDI4kCQ+8MwAtJUyZVCkaRnoAEohyigGRraQsGmYbbV9iHToAFKsRqnpAThFJRPMPkQA5ujMwyERMYEOrEXQDYXFpYlqPswAFiZgZkF8tvhBDlQCPmpqQjiQTk9gAvhNRgBRcj7VbXCZWHysdAAVSIiGGVBOPB8iEQygA7vMrutEgAzZREeDoADCRlY0DOtGYNKRcG22QAYpp0AA5HykEIkm7mRn9e7oVBLLrPNYi04Aa2gagAMtYaQ90PAFcKIcoUQhNdrJdkMBRJQBBeDU1iy64EcEESBsRqjDC2V48FTqHm6fSGYwe1ztTorJSqDqwSpAiCgh0gUFAA).

Copied here for convenience:

```ts
declare const actionData: {
    email: string | string[];
    password: string | string[];
} | undefined

export const result = actionData?.email?.[0];
                                    // ^^ -> ts-eslint thinks this is unnecessary
```

I have TS `noUncheckedIndexedAccess` option turned on in my project so this should not be unnecessary, and it causes an error in production like so: `Cannot read properties of undefined (reading '0')`

**Comment by bradzacher** [comment]
@frontsideair you can access string characters using array index notation. So your optional chain does nothing - the report is correct.

You need to write stricter code using either `typeof` or `Array.isArray`.

**Comment by frontsideair** [comment]
First of all, thanks for taking the time to look into this.

However I would never expect ts-eslint to work with the types I didn't provide, but the type I provided is not `string | string[]`, it's `string | string[] | undefined`, as can be seen from the screenshot.

Also I removed the `string` from the union and it was still reported as unnecessary, [here's the playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3lMGCWqAIlBlAFzwDeAUPHfCALZR4QWYx4oDmA2gLoBuWvQAOUNGgDuSGMHYZOPAcIC+8AD7wArilAAzLiGDVqIAB6jZWZOixw02iFgC82XARTFSAfgB0TCwQ-rwADEJAA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCUAAgE7QC2AlrALTECuSA9NGgDYVLzUDGA9kgDMKAc0gAafBEglyVWg2ZsOXPoJGMkvFNHGSoAB1Lx4FaMV0EDrOsI6ZC8AJ76W3YhX1cW7To1J8yMmRtFAsCSH1rWyR7Jxc0Nw8vJV9-XkDg6BRaaABHOgp3JGFqOOgeAAtobgBrDlEJSwiouwdnV3dPam9lRjR4d254SHwAXUaDAENiNDMsIjKEzuSfeEZ9adnzCYjNswB5Twp+DGxQJsNeACtq4ew8SykAOkZ4BP4hYSertH4Ry1G+AAvjtIjYOKdcHppItEl0epx-mBxvhIMReLw7mABnRoDt6KwWPNzuFNNQAO7TJD1HjpIKcSEPR4wYjo7Z6CAkx5QRDEMiMjnMoSwIJI7ko7lSVi8biTEx-bCQSZIRzkqqkMUEIF6CXhNrxOErXpkgBGk1mpV41H6RVEirMbLCUn1HSS3RSa1I+UK5UmlIoWMglOISCdC3aSzdCLWDBNvAY2SC8AqWnmkF4AgEYZhEcN7tWGityfR5Oo7F5k1YaYzWYmzthy3zvVUaAo-WQXDK1G0Qmp8qQAu5QapYakZU1uvrucb0cL1AYSGqLDQ00ctNQAeOocVwdDwJ2Ik0pAACnLeQP5kypIXtAB9MhaOiEjB1qDMABMOlfkEYk30FGzRgTQKVhQm-dY6BNdhuCAkCwOhAAqBDGAQ74MDGEAgSAA&tsconfig=N4KABGBECWB2DGAbArgEwKaQFxgNrgigCd0BbaADwDp1YA3K1KgFwGdIAaAiSAKl4D0vFuy6Eo-ISIqQCAXTFR4Ae1IAHaInREA8mubRlsdjlDjIiaACNseboUipVnezyekq0ZtoCGVrS7iPOisAEwADACMAJyy4gr2kCEAssqoyFoAkrDeRMpqtsxEyOiKPABWrDI4kCQ+8MwAtJUyZVCkaRnoAEohyigGRraQsGmYbbV9iHToAFKsRqnpAThFJRPMPkQA5ujMwyERMYEOrEXQDYXFpYlqPswAFiZgZkF8tvhBDlQCPmpqQjiQTk9gAvhNRgBRcj7VbXCZWHysdAAVSIiGGVBOPB8iEQygA7vMrutEgAzZREeDoADCRlY0DOtGYNKRcG22QAYpp0AA5HykEIkm7mRn9e7oVBLLrPNYi04Aa2gagAMtYaQ90PAFcKIcoUQhNdrJdkMBRJQBBeDU1iy64EcEESBsRqjDC2V48FTqHm6fSGYwe1ztTorJSqDqwSpAiCgh0gUFAA).

I would expect this rule to not report the last optional chain as unnecessary.

<img width="443" alt="Screen Shot 2022-10-15 at 19 55 17" src="https://user-images.githubusercontent.com/868283/195998611-32d57609-f3bb-47f6-a28d-a7a7cf96fb31.png">

**Comment by bradzacher** [comment]
@frontsideair that's not how optional chaining works!

Based on your types `actionData?.email[0]` may return undefined if and only if `actionData` is null or undefined.
Optional chaining is a short-circuiting operator. That means that it **_stops_** evaluating the chain if it's undefined.

Put another way, it's equivalent to this code:

```ts
// actionData?.email[0] is the same as
let result = undefined;
if (actionData != null) {
  result = actionData.email[0];
}
```

Note how the `.email` and `[0]` is only evaluated if and only if the chained object is non nullish?

Whilst the result of `actionData?.email` may be undefined, that is not because the email property may be undefined. If you inspect the return type of `actionData?.email[0]` you'll note that it's also possibly undefined. Again this is because of the first optional chain.

To further illustrate, if you were to write your ***unnecessary*** code (based on your types), you would end up with this equivalent code:

```ts
// actionData?.email?.[0] is the same as
let result = undefined;
if (actionData != null) {
  if (actionData.email != null) {
    result = actionData.email[0];
  }
}
```

See now how that second optional chain is truly unnecessary based on your types?
The rule is correctly flagging your code based on your declared types.

**Comment by frontsideair** [comment]
Yes, you are absolutely correct. I was confused because the types I'm actually working were coming from a library and it seems to be wrong. I simplified them as they were shown and got it wrong by extension. I'll try to open an issue with the library itself, thanks for clarifying!

**Comment by JoshuaKGoldberg** [comment]
Good news: at some point, the OP's sample was fixed! https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAZiEMC8MDaByAhhgNDDAIzwOAwF0BuGAehpigE8AHAUwC4ZoAnASzADmacgChQkWISzcU8RGgCsVWvSZtOPfgPwA3LABsArhxiGwAE1Zx%2BrcyJF0YABSwQIrCAwAWrGN0P6rCK8cDAAFAggiuQAlDAA3iIwMOIQIIEAdPogAhHySjGUIgC%2B9o4AYli8%2Bp5QPn4BQSHhUtxxicmp6axZOWGthSVAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Brad's first response also works: https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPJjAJYoDmA2gLoBQty6WARrPALw56UCs1A3LRLZ4AClYwAlPADe8AL60gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

An enterprising individual could go back through https://github.com/typescript-eslint/typescript-eslint/pulls?q=no-unnecessary-condition+in%3Atitle+is%3Amerged to figure out when this was fixed. But even without that, I think this issue is closed. Cheers!
